log/operation_manager.go

```diff
@@ -256,25 +251,42 @@ func (o *OperationManager) masterFor(ctx context.Context, allIDs []int64) ([]int
 	return heldIDs, nil
 }
 
-func (o *OperationManager) runElection(ctx context.Context, logID string) (context.CancelFunc, error) {
+// runElectionWithRestarts runs the election/resignation loop for the given log
+// indefinitely, until the returned CancelFunc is invoked. Any failure during
+// the loop leads to a restart of the loop with a few seconds delay.
+func (o *OperationManager) runElectionWithRestarts(ctx context.Context, logID string) context.CancelFunc {
 	glog.Infof("create master election goroutine for %v", logID)
 	cctx, cancel := context.WithCancel(ctx)
-	e, err := o.info.Registry.ElectionFactory.NewElection(cctx, logID)
-	if err != nil {
-		cancel()
-		return nil, fmt.Errorf("failed to create election for %v: %v", logID, err)
-	}
-	o.runnerWG.Add(1)
-	go func() {
-		defer o.runnerWG.Done()
+	run := func(ctx context.Context) {
+		e, err := o.info.Registry.ElectionFactory.NewElection(ctx, logID)
```

Comment:
Previously if the election was fundamentally broken these goroutines would return an error, now they'll loop.

We might need to look at adding metrics for election restart rate.

Reply:
Agree.

I mitigated this in Migrillian by having a metric of restarts as you suggest, but instead of restarting only the election loop, I restarted the whole "operation" for this log. I think it will be a better granularity.

This will be easy to do if we restructure log operation as suggested in #1640, so that all the logs are nicely isolated (and only share some semaphores limiting the number of simultaneously running operations).

I'll add a TODO here.